File: split/package.pas
procedure package(c: smallnumber);
var
  h: scaled;
  p: halfword;
  d: scaled;
begin
  d := eqtb[5837].int;
  unsave;
  saveptr := saveptr - 3;
  if curlist.modefield = -102 then
    curbox := hpack(mem[curlist.headfield].hh.rh, savestack[saveptr + 2].int, savestack[saveptr + 1].int)
  else
  begin
    curbox :=
      vpackage(mem[curlist.headfield].hh.rh, savestack[saveptr + 2].int, savestack[saveptr + 1].int, d);
    if c = 4 then{1087:}

    begin
      h := 0;
      p := mem[curbox + 5].hh.rh;
      if p <> 0 then
        if mem[p].hh.b0 <= 2 then
          h := mem[p + 3].int;
      mem[curbox + 2].int := mem[curbox + 2].int - h + mem[curbox + 3].int;
      mem[curbox + 3].int := h;
    end{:1087}
;
  end;
  popnest;
  boxend(savestack[saveptr + 0].int);
end;

File: cpp/src/package.cpp
#include "package.h"
#include "sauvegarde.h"
#include "popnest.h"
#include "boite.h"

static int box_max_depth(void) { return dimen_par(box_max_depth_code); }

void package(smallnumber c, Token t)
{
	unsave();
	auto s2 = savestack.back().int_;
	savestack.pop_back();
	auto s1 = savestack.back().int_;
	savestack.pop_back();
	auto s0 = savestack.back().int_;
	savestack.pop_back();
	if (mode == -hmode)
		curbox = hpack(head->link, s2, s1);
	else
	{
		curbox = vpackage(head->link, s2, s1, box_max_depth());
		if (c == 4)
		{
			scaled h = 0;
			auto p = curbox->list_ptr;
			if (p && p->type <= rule_node)
				h = dynamic_cast<RuleNode*>(p)->height;
			curbox->depth += -h+curbox->height;
			curbox->height = h;
		}
	}
	popnest();
	boxend(s0);
}
